File: src/main/java/com/contentvelocity/service/ScoreCalculator.java
package com.contentvelocity.service;

import com.contentvelocity.model.AiScores;
import com.contentvelocity.model.AuditCheck;
import com.contentvelocity.model.PageData;
import com.contentvelocity.model.PageScores;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
public class ScoreCalculator {

    public PageScores calculate(List<AuditCheck> checks, AiScores ai, PageData page) {

        // SEO score: rule-based 75% + AI boost 25% (unchanged)
        int seoEarned = 0, seoMax = 0;
        for (AuditCheck c : checks) {
            if (isSeoCategory(c.category)) {
                seoMax += c.weight;
                if (c.pass) seoEarned += c.weight;
            }
        }
        int seoBase  = seoMax > 0 ? Math.round((float) seoEarned / seoMax * 75) : 0;
        int seoBoost = ai != null ? calcSeoAiBoost(ai) : 10;
        int seoScore = Math.min(100, seoBase + seoBoost);

        // GEO score: geo-rules formula (50% objective + 50% subjective)
        double objectiveScore  = calcObjectiveGeoScore(page);
        double subjectiveScore = ai != null ? calcSubjectiveGeoScore(ai) : 0.5;
        int geoScore = (int) Math.round((0.5 * objectiveScore + 0.5 * subjectiveScore) * 100);
        geoScore = Math.min(100, Math.max(0, geoScore));

        return new PageScores(seoScore, geoScore);
    }

    private boolean isSeoCategory(String category) {
        return category.equals("On-page") ||
               category.equals("Technical") ||
               category.equals("Content");
    }

    private int calcSeoAiBoost(AiScores ai) {
        // clarity ≈ readability (0-10), relevance ≈ intent match (0-8), authority ≈ depth (0-7)
        return (int) Math.round(ai.clarity * 10 + ai.relevance * 8 + ai.authority * 7);
    }

    private double calcObjectiveGeoScore(PageData p) {
        // structuredData (20%)
        double structuredData = p.totalSchemaFields == 0 ? 0.0
            : clamp((double) p.validSchemaFields / p.totalSchemaFields);

        // citationReadiness (20%)
        double citationReadiness = clamp(
            0.4 * clamp(p.externalLinkCount / 10.0) +
            0.3 * clamp(p.internalLinkCount  / 10.0) +
            0.3 * clamp(p.quotationCount     / 5.0));

        // factualDensity (20%)
        double factualDensity = p.totalSentences == 0 ? 0.0
            : clamp((double) p.factualStatementCount / p.totalSentences);

        // contentModularity (15%): clamp(50/avgChunkSize) per geo_ruling.js
        double chunkScore = p.avgChunkSize == 0 ? 0.0 : clamp(50.0 / p.avgChunkSize);
        double contentModularity = clamp(
            0.4 * clamp(p.headingCount / 10.0) +
            0.3 * clamp(p.listCount    / 10.0) +
            0.3 * chunkScore);

        // technicalDensity (15%): normalise ratio against 10% baseline per geo_ruling.js
        double technicalDensity = p.totalWords == 0 ? 0.0
            : clamp(((double) p.technicalTermCount / p.totalWords) / 0.1);

        // freshness (10%)
        double freshness = p.daysSinceLastUpdate == null ? 0.4
            : p.daysSinceLastUpdate < 30  ? 1.0
            : p.daysSinceLastUpdate < 180 ? 0.7 : 0.4;

        return 0.20 * structuredData +
               0.20 * citationReadiness +
               0.20 * factualDensity +
               0.15 * contentModularity +
               0.15 * technicalDensity +
               0.10 * freshness;
    }

    private double calcSubjectiveGeoScore(AiScores ai) {
        return 0.20 * ai.relevance +
               0.15 * ai.authority +
               0.15 * ai.clarity +
               0.15 * ai.conversationalFit +
               0.10 * ai.uniqueness +
               0.10 * ai.engagement +
               0.15 * ai.trustworthiness;
    }

    private double clamp(double v) {
        return Math.max(0.0, Math.min(1.0, v));
    }
}

File: src/services/scoreCalculator.js
const SEO_CATEGORIES = new Set(["On-page", "Technical", "Content"]);

function clamp(v) {
    return Math.max(0.0, Math.min(1.0, v));
}

function band(score) {
    if (score >= 80) return "Excellent";
    if (score >= 60) return "Good";
    if (score >= 40) return "Needs work";
    return "Critical";
}

function calcSeoAiBoost(ai) {
    return Math.round(ai.clarity * 10 + ai.relevance * 8 + ai.authority * 7);
}

function calcObjectiveGeoScore(page) {
    // structuredData (20%)
    const structuredData =
        page.totalSchemaFields === 0
            ? 0.0
            : clamp(page.validSchemaFields / page.totalSchemaFields);

    // citationReadiness (20%)
    const citationReadiness = clamp(
        0.4 * clamp(page.externalLinkCount / 10.0) +
            0.3 * clamp(page.internalLinkCount / 10.0) +
            0.3 * clamp(page.quotationCount / 5.0),
    );

    // factualDensity (20%)
    const factualDensity =
        page.totalSentences === 0
            ? 0.0
            : clamp(page.factualStatementCount / page.totalSentences);

    // contentModularity (15%)
    const chunkScore =
        page.avgChunkSize === 0 ? 0.0 : clamp(50.0 / page.avgChunkSize);
    const contentModularity = clamp(
        0.4 * clamp(page.headingCount / 10.0) +
            0.3 * clamp(page.listCount / 10.0) +
            0.3 * chunkScore,
    );

    // technicalDensity (15%)
    const technicalDensity =
        page.totalWords === 0
            ? 0.0
            : clamp(page.technicalTermCount / page.totalWords / 0.1);

    // freshness (10%)
    let freshness = 0.4; 

    if (page.daysSinceLastUpdate != null && page.daysSinceLastUpdate < 30)
        freshness = 1.0;
    else if (page.daysSinceLastUpdate != null && page.daysSinceLastUpdate < 180)
        freshness = 0.7;

    return (
        0.2 * structuredData +
        0.2 * citationReadiness +
        0.2 * factualDensity +
        0.15 * contentModularity +
        0.15 * technicalDensity +
        0.1 * freshness
    );
}

function calcSubjectiveGeoScore(ai) {
    return (
        0.2 * ai.relevance +
        0.15 * ai.authority +
        0.15 * ai.clarity +
        0.15 * (ai.conversationalFit || ai.conversational_fit || 0) +
        0.1 * ai.uniqueness +
        0.1 * ai.engagement +
        0.15 * ai.trustworthiness
    );
}

function calculate(checks, aiScores, page) {
    // SEO score: rule-based 75% + AI boost 25%
    let seoEarned = 0,
        seoMax = 0;
    for (const c of checks) {
        if (SEO_CATEGORIES.has(c.category)) {
            seoMax += c.weight;
            if (c.pass) seoEarned += c.weight;
        }
    }
    const seoBase = seoMax > 0 ? Math.round((seoEarned / seoMax) * 75) : 0;
    const seoBoost = aiScores ? calcSeoAiBoost(aiScores) : 10;
    const seo = Math.min(100, seoBase + seoBoost);

    // GEO score: 50% objective + 50% subjective
    const objectiveScore = calcObjectiveGeoScore(page);
    const subjectiveScore = aiScores ? calcSubjectiveGeoScore(aiScores) : 0.5;
    const geo = Math.min(
        100,
        Math.max(
            0,
            Math.round((0.5 * objectiveScore + 0.5 * subjectiveScore) * 100),
        ),
    );

    const combined = Math.round(seo * 0.5 + geo * 0.5);

    return {
        seo,
        geo,
        combined,
        seoBand: band(seo),
        geoBand: band(geo),
        combinedBand: band(combined),
    };
}

module.exports = { calculate };
